### PR TITLE
Add schema-derived types for editable tree 2 proxies

### PR DESCRIPTION
**Pull request by noencke**
## Description

This creates adaptations of the editable-tree-2 schema-derived types which:

1. Produce the "proxy" API objects
2. "Unbox" unconditionally

It also renames "List" to "SharedTreeList".

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts [hunk 1]
```diff
@@ -3,16 +3,33 @@
...
 	CheckTypesOverlap,
 	FlexibleNodeContent,
 	Sequence,
+	NodeKeyField,
+	AssignableFieldKinds,
 } from "../editableTreeTypes";
 
 /** Implements 'readonly T[]' and the list mutation APIs. */
-export interface List<TTypes extends AllowedTypes> extends ReadonlyArray<UnboxNodeUnion<TTypes>> {
+export interface SharedTreeList<TTypes extends AllowedTypes>
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:31]: Why the rename? Do we plan on doing the same thing with map?

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts [hunk 2]
```diff
@@ -131,3 +148,98 @@ export interface List<TTypes extends AllowedTypes> extends ReadonlyArray<UnboxNo
...
+export type SharedTreeObject<TSchema extends StructSchema> = ObjectFields<
+	TSchema["structFieldsObject"]
+>;
+
+/**
+ * Helper for generating the properties of a {@link SharedTreeObject}.
+ * @alpha
+ */
+export type ObjectFields<TFields extends RestrictiveReadonlyRecord<string, FieldSchema>> = {
+	// Add getter only (make property readonly) when the field is **not** of a kind that has a logical set operation.
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:165]: Is this still true given the new typing?

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts (hunk 1 above)
Comment (taylorsw04) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:31]: I see we did. Out of curiosity, why?

Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:31]: Yes. `SharedTreeObject` and `SharedTreeMap` are just below. "List" is okay, but we can't call the object "Object" and the map "Map".

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts (hunk 2 above)
Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:165]: Right now this means we won't allow the setting of lists (and of node identifiers). I actually can't remember right now whether we said we can set lists or not. `myNode.list = [a, b, c]`, as opposed to having to delete all elements and then add new elements. I think we don't allow the setting, but I could have forgotten. Do you remember?

Comment (taylorsw04) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:165]: right, this requires slice delete. nvm.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts [hunk 3]
```diff
@@ -131,3 +148,98 @@ export interface List<TTypes extends AllowedTypes> extends ReadonlyArray<UnboxNo
...
+	Kind extends typeof FieldKinds.nodeKey
+	? NodeKeyField
+	: // TODO: forbidden
+	  unknown;
+
+/**
+ * Given multiple node schema types, return the corresponding object type union in the proxy-based API.
+ * @alpha
+ */
+export type ProxyNodeUnion<TTypes extends AllowedTypes> = TTypes extends readonly [Any]
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:220]: can you doc how this type works?

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+			assert.equal(root.polyValueChild, 42);
+		} else {
+			assert.equal(root.polyValueChild.content, 42);
+		}
+	});
+});
+
+function createTypedTreeView<TRoot extends FieldSchema>(
+	schema: TypedSchemaCollection<TRoot>,
+	initialTree: any,
```
Comment (taylorsw04) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts:116]: type this plz

**Review by taylorsw04 — APPROVED** [review]

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts (hunk 4 above)
Comment (noencke) [reply experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts:116]: I'll do this in a followup PR.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts (hunk 3 above)
Comment (noencke) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/types.ts:219]: You bet.

**Pull request merged**